### PR TITLE
Refactor DeleteSourceDialog so it accepts Set[Source]

### PR DESCRIPTION
**Pull request by rocodes**
## Status

Ready for review

## Description

Refactor DeleteSourceDialog so it takes a set of sources instead of a single source. Display a user-friendly error message if the empty set is provided. Add parameterized unit tests to test empty, n=1, and n>1 conditions. This does not include any actual batch delete functionality, but it is in preparation for that change.

I was tempted to also refactor the DeleteSource and DeleteConversation dialogs into one dialog element, since there's no reason not to and is in line with what we did on the Print, Print Conversation, and Print Transcript dialogs, but I think it's too much scope creep. But it would be a straightforward refactor PR later on if anyone wants to take it on.

This PR makes minor string changes that will need il8n.

Towards #2160 

## Test Plan

- [ ] Visual review
- [ ] CI

## Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [x] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration is [self-contained] and applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

[self-contained]: https://github.com/freedomofpress/securedrop-client#generating-and-running-database-migrations

### CODE REVIEW

**Comment by rocodes** [comment]
(realized I want to add one more test before ready for review :) )

File: client/securedrop_client/gui/source/delete/dialog.py [hunk 1]
```diff
@@ -58,4 +68,23 @@ def make_body_text(self) -> str:
             "<p>&nbsp;</p>",
         )
 
-        return "".join(message_tuple).format(source=f"<b>{self.source.journalist_designation}</b>")
+        return "".join(message_tuple).format(
+            source=f"<b>{self._get_source_names(self.sources)}</b>"
```
[the remark this answers is not in the document]
Comment (rocodes) [reply client/securedrop_client/gui/source/delete/dialog.py]: oof. fixed in https://github.com/freedomofpress/securedrop-client/pull/2188/commits/a0e9d56034f1e4de827ebab5d5af98640c8e89a4

File: client/securedrop_client/gui/source/delete/dialog.py [hunk 2]
```diff
@@ -26,17 +26,27 @@
 class DeleteSourceDialog(ModalDialog):
```
[the remark this answers is not in the document]
Comment (rocodes) [reply client/securedrop_client/gui/source/delete/dialog.py:26]: This is the only change I haven't made - I think it's marginally more readable with the singular form. With ExportDialog, PrintDialog, we get around it cause there is no noun, but imo docstrings, comments, type hinting should make it clear.

**Comment by rocodes** [comment]
> ### Alternative approach suggestion
> 
> This is one of those situations where some UX help would be great 🙂

Hey, thanks for your thoughtful comment. I do agree that ux help would be beneficial. Here're my initial thoughts; I'm open to discussion.

There are a few known UX patterns we can fall back on, specifically "Confirmation Modal for Destructive Operation", "Bulk Action best practices", and "Actions/affordances in a table like/scrolling list view".  (These might not be exact names but give a decent idea of what to search for. Links are below.)

Confirmation modal for destructive operations, best practices are:
- [x] Dangerous option should be a different colour and not the default (so default is Cancel, not Delete)
- [x] Dangerous option should explain precisely what's happening ("Are you sure you want to delete the following sources" instead of "Are you sure you want to do this/delete 3 sources?")
- [ ] :x: Super dangerous actions should require more than a button click (eg github typing the project name to confirm deleting the project), but doing this too often frustrates users
- [ ] :x: Consider Progressive Disclosure ("See more ...." / details of action hidden but visible before deletion)
- [ ] :x: Consider "Don't ask me again" to dismiss confirmation dialog

Bulk action best practices:
- [x] (this is the plan- checkbox) Clear affordance (if items have a checkbox, you know they can be checked/selected) 
- [x] (this is the plan) Visual indicator of which items are selected
- [ ]  :x: Usually, bulk actions are at the top or bottom of the table/row of items they will act on (think about the Gmail bar)

Destructive action best practices:
- [x] Don't put it near a nondestructive or routine action, to avoid the risk of mis-clicks
- [x] Launch a confirmation modal per above
- [x] Don't overuse destructive action confirmations to avoid user fatigue 

IMO repeated confirmation dialogs would not be better than the current DeleteSource implementation; the goal of batch operations is to save the user clicks. (This is also why I am advocating for a sidebar and not a dropdown menu for the Batch delete action). I would also worry about encouraging users to click through Destructive dialogs quickly in order to complete a task, so I am in favour of one dialog for a batch operation, which is in line with batch operations ux practices generally. But, it is a destructive operation, so it definitely does need a level of confirmation.

The areas where I feel we are departing from UX norms that should be discussed are:
- Batch action is not at the top of SourceConversationView, but is on a side panel. This is open for debate and was mostly me trying to a) ensure destructive/global actions are in a different place than other actions (avoid misclicks, avoid crowded ui areas). Traditionally, this would go at the top of the sourcelistwidget.
- We don't offer an Undo button _and_ we aren't requiring an unusual action ("type "delete" to proceed").  I think we should not change anything on this front because of the frequency of this operation as an anti-spam measure.
- We aren't doing progressive disclosure (previews of messages). I feel like this would make the feature much more complex, but could be a possible reasonable enhancement. 

(Links - based on a lot from nngroup, who are longstanding/practitioners in/creators of the ux field, see eg from them and others:
https://www.nngroup.com/articles/confirmation-dialog/
https://uxpsychology.substack.com/p/how-to-design-better-destructive
https://ux.folio.org/docs/guidelines/ux-patterns/using-the-confirmation-modal/
https://www.nngroup.com/articles/progressive-disclosure/
https://www.nngroup.com/articles/data-tables/ (The section on Batch Actions, although it's brief)
https://www.nngroup.com/articles/closeness-of-actions-and-objects-gui/ (where should the actions go?)

**Comment by rocodes** [comment]
> The following sources have had replies from journalists: source-one, source-two 

I think this is a great idea in terms of differentiating what might be spam, and from an implementation perspective, it would not be very complex, but I think we should discuss if it's a must-do at this stage, or an enhancement. It also introduces a new UX condition (so would it be present for all deletion operations including single-source deletion, or just multi-source deletion?).

File: client/securedrop_client/gui/source/delete/dialog.py [hunk 3]
```diff
@@ -26,23 +26,36 @@
...
+        self.sources = sources
 
-        self.source = source
+        # If the dialog is constructed with no sources, show a warning; otherwise,
+        # confirm the number and designation of the sources to be deleted
+        count = len(sources)
+        if count == 0:
+            self._show_warning_nothing_selected()
+        else:
+            if count == 1:
```
Comment (legoktm) [inline client/securedrop_client/gui/source/delete/dialog.py:39]: This should use ngettext instead of manually implementing the plural.

**Review by legoktm — COMMENTED** [review]
I don't have an opinion on the UX question right now, I do think that if we're stuck or uncertain, sticking close to what the Guardian did is fine/what I'd recommend and then working on improvements afterwards.

File: client/securedrop_client/gui/source/delete/dialog.py [hunk 4]
```diff
@@ -58,4 +71,23 @@ def make_body_text(self) -> str:
             "<p>&nbsp;</p>",
         )
 
-        return "".join(message_tuple).format(source=f"<b>{self.source.journalist_designation}</b>")
+        return "".join(message_tuple).format(
+            source_or_sources=f"<b>{self._get_source_names(sources)}</b>"
```
[the remark this answers is not in the document]
Comment (rocodes) [reply client/securedrop_client/gui/source/delete/dialog.py:75]: Hm, thanks. I will figure out the styling element and add a test of some sort to check for this.

[the remark this answers is not in the document]
Comment (rocodes) [reply client/securedrop_client/gui/source/delete/dialog.py:75]: For our future selves: https://doc.qt.io/qt-5/layout.html#layout-issues  (Even if WordWrap is True, there's more to do.)

**Comment by cfm** [comment]
On Thu, Sep 05, 2024 at 09:27:31AM -0700, Kunal Mehta wrote:
> I don't have an opinion on the UX question right now, I do think that
> if we're stuck or uncertain, sticking close to what the Guardian did
> is fine/what I'd recommend and then working on improvements
> afterwards.

I second this.

**Comment by rocodes** [comment]
Thanks all for the review + comments. I've fixed the source designator display issue (see below) and also added a regression test to catch that test case. 
delete_multi

File: client/securedrop_client/gui/source/delete/dialog.py (hunk 3 above)
Comment (rocodes) [reply client/securedrop_client/gui/source/delete/dialog.py]: Thank you , resolved in https://github.com/freedomofpress/securedrop-client/pull/2188/commits/16e48a80d1d74758c5503d4c48c851aa5db5fb63

**Comment by rocodes** [comment]
Marking again as ready for review.
- I've given up on the regression test for the dialog text cutoff size.  However, in all of my manual testing, the text is resized correctly. https://github.com/freedomofpress/securedrop-client/pull/2188#discussion_r1747349928  
- I've addressed other review feedback.

File: client/tests/gui/source/delete/test_dialog.py [hunk 5]
```diff
@@ -1,22 +1,109 @@
...
+        assert dialog.text() == "No sources have been selected."
+
+    def test_no_sources_continue_button_disabled(self, dialog):
+        if len(dialog.sources) > 0:
+            pytest.skip("Skip if sources")
+
+        assert not dialog.continue_button.isEnabled()
+
+    @pytest.mark.parametrize(
+        "names", [["source one"], ["source one", "source two", "source three"]]
```
[the remark this answers is not in the document]
Comment (rocodes) [reply client/tests/gui/source/delete/test_dialog.py]: Thanks, addressed in https://github.com/freedomofpress/securedrop-client/pull/2188/commits/b128e48cc04d7daa487b6c5890e38644872852cc (The dialog parameterization is intentional but not needed for this test)

File: client/tests/gui/source/delete/test_dialog.py [hunk 6]
```diff
@@ -1,22 +1,109 @@
...
-    def test_displays_important_information_when_shown(self):
-        assert "not be able to send them replies" in self.dialog.text()
-        assert "source will not be able to log in" in self.dialog.text()
-        assert "files and messages from that source will also be destroyed" in self.dialog.text()
+        dialog = DeleteSourceDialog(sources)
+        font_metrics = dialog.body.fontMetrics()
+        text_rect = font_metrics.tightBoundingRect(dialog.body.text())
+        text_area = text_rect.height() * text_rect.width()
+        body_area = dialog.body.width() * dialog.body.height()
+        assert text_area <= body_area
```
[the remark this answers is not in the document]
Comment (legoktm) [reply client/tests/gui/source/delete/test_dialog.py]: I think testing the word wrapping is reasonable to tackle as a GUI test instead (once we have those).

File: client/tests/gui/source/delete/test_dialog.py [hunk 7]
```diff
@@ -1,22 +1,91 @@
...
+    conditions.
+    """
+    # Give the source(s) a submission
+    for source in request.param:
+        factory.File(source=source)
+    return DeleteSourceDialog(request.param)
 
-    def test_default_button_is_safer_choice(self):
+
+class TestDeleteSourceDialog:
```
Comment (legoktm) [inline client/tests/gui/source/delete/test_dialog.py:22]: I appreciate the switch from unittest to pytest. Now that dialog is a fixture and not a member variable, there's no need for a class since `self` is never used AFAICT unless you find it useful for organization (I personally don't).

File: client/tests/gui/source/delete/test_dialog.py [hunk 8]
```diff
@@ -1,22 +1,91 @@
...
+            source = factory.Source(journalist_designation=item)
+            sources.update([source])
+
+        dialog = DeleteSourceDialog(sources)
 
-    def test_displays_important_information_when_shown(self):
-        assert "not be able to send them replies" in self.dialog.text()
-        assert "source will not be able to log in" in self.dialog.text()
-        assert "files and messages from that source will also be destroyed" in self.dialog.text()
+        assert all(n in dialog.make_body_text(sources) for n in names), dialog.make_body_text(
```
Comment (legoktm) [inline client/tests/gui/source/delete/test_dialog.py:89]: I would prefer if we avoid using `all` in tests because it makes it harder to debug failures ("Failed asserting True is False", etc.). Unrolling it wouldn't be too bad IMO:
```python
        text = dialog.make_body_text(sources)
        for name in names:
            assert name in text
```

**Review by legoktm — APPROVED** [review]
Thanks, LGTM. I did a full review since deeplow's last review. The comments are just personal preference things in tests and, if desired, can be addressed in follow-ups.

File: client/tests/gui/source/delete/test_dialog.py (hunk 7 above)
Comment (rocodes) [reply client/tests/gui/source/delete/test_dialog.py:22]: I do honestly like it for organization and consistency with other unittests (and for ease of doing things like parameterization on the whole test class if wanted) and it's a style I default to, but ... I am open to other thoughts :)

File: client/tests/gui/source/delete/test_dialog.py (hunk 8 above)
Comment (rocodes) [reply client/tests/gui/source/delete/test_dialog.py]: thanks, addressed in https://github.com/freedomofpress/securedrop-client/pull/2188/commits/1e541a6694678839c20305ff9084bd3f7e0f7cf6

**Review by cfm — APPROVED** [review]
Approving the test-only change in <https://github.com/freedomofpress/securedrop-client/compare/b128e48cc04d7daa487b6c5890e38644872852cc..1e541a6694678839c20305ff9084bd3f7e0f7cf6> after the full review in <https://github.com/freedomofpress/securedrop-client/pull/2188#pullrequestreview-2287274646>.
